sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Serialization/SchemaRegistrySerializer.cs

```diff
@@ -0,0 +1,570 @@
+        {
+            Argument.AssertNotNull(data, nameof(data));
+            Argument.AssertNotNull(contentType, nameof(contentType));
+
+            string[] contentTypeArray = contentType.ToString().Split('+');
+            if (contentTypeArray.Length != 2 || contentTypeArray[0] != _mimeType)
+            {
+                throw new FormatException($"Content type was not in the expected format of '{_mimeType}+schema-id', where 'schema-id' " +
+                                          "is the Schema Registry schema ID.");
+            }
+
+            string schemaId = contentTypeArray[1];
+
+            if (async)
+            {
+                return await DeserializeInternalAsync(data, dataType, schemaId, true, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                return DeserializeInternalAsync(data, dataType, schemaId, false, cancellationToken).EnsureCompleted();
+            }
+        }
+
+        private async ValueTask<object> DeserializeInternalAsync(
+            BinaryData data,
+            Type dataType,
+            string schemaId,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            Argument.AssertNotNull(schemaId, nameof(schemaId));
+
+            string schemaDefinition;
+            try
+            {
+                // Attempt to get the schema definition using the schema Id from the Content Type
+                if (async)
+                {
+                    schemaDefinition = await GetSchemaByIdAsync(schemaId, true, cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    schemaDefinition = GetSchemaByIdAsync(schemaId, false, cancellationToken).EnsureCompleted();
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new Exception($"An error occurred while attempting to retrieve the schema with id {schemaId}.", ex);
+            }
+
+            var deserializer = _serializerOptions.Serializer;
+            object objectToReturn;
+            try
+            {
+                // Attempt to deserialize
+                var dataStream = data.ToStream();
+                if (async)
+                {
+                    objectToReturn = await deserializer.DeserializeAsync(dataStream, dataType, cancellationToken: cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    objectToReturn = deserializer.Deserialize(dataStream, dataType, cancellationToken);
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new Exception($"An error occurred while attempting to deserialize the data.", ex);
+            }
+
+            try
+            {
+                // Attempt to validate
+                _schemaValidator.Validate(objectToReturn, dataType, schemaDefinition);
```

Comment:
I thought Validate will already throw an AggregateException if there are errors returned, so why do we need to wrap again in another Exception?

Reply:
Hmm I guess Validate is virtual so users can override it. Still it seems excessive to have Exception -> AggregateException -> UserException for the common case given that the user is supplying the Validator and has complete control over what gets thrown. My vote would be to not add this last layer of wrapping and to push the message down into our Validate implementation.

Reply:
Yeah that sounds good. I had it there for consistency, since all the other exceptions being thrown in this method are wrapped in an exception / I think this is what aligns with the Avro serializer

Reply:
But I agree that having that many layers of exception is unnecessary